cf-net/cf-net.c

```diff
@@ -0,0 +1,766 @@
+#include <communication.h>
+#include <rlist.h>
+#include <scope.h>
+#include <policy.h>
+#include <audit.h>
+#include <connection_info.h>
+#include <loading.h>
+#include <expand.h>
+#include <files_hashes.h>
+*/
+
+typedef struct
+{
+    bool debug;
+    bool verbose;
+    bool inform;
+    bool used_default;
+} CFNetOptions;
+
+// INIT:
+static void CFNetSetDefault(CFNetOptions *opts);
+static void CFNetInit();
+
+// MAIN LOGIC:
+static int CFNetParse(int argc, char **argv,                        // INPUTS
+               CFNetOptions *opts, char ***args, char **hostnames); // OUTPUTS
+static int CFNetCommandSwitch(CFNetOptions *opts, const char *hostname, char **args);
+static int CFNetRun(CFNetOptions *opts, char **args, char *hostnames);
+static char *RequireHostname(char *hostnames);
+
+// PROTOCOL:
+static AgentConnection* CFNetOpenConnection(const char *server);;
+static int JustConnect(const char *server, char* port);
+
+// COMMANDS:
+static int CFNetHelpTopic(const char *topic);
+static int CFNetHelp(const char *topic);
+static int CFNetConnectSingle(const char *server, bool print);
+static int CFNetConnect(CFNetOptions *opts, const char* hostname, char **args);
+static void CFNetStatPrint(const char* file, int st_mode, const char *server);
+static int CFNetStat(CFNetOptions *opts, const char* hostname, char **args);
+static int CFNetGet(CFNetOptions *opts, const char* hostname, char **args);
+static int CFNetOpenDir(CFNetOptions *opts, const char* hostname, char **args);
+static int CFNetMulti(const char *server);
+static int CFNetMultiTLS(const char *server);
+
+static const char *const CF_NET_SHORT_DESCRIPTION =
+    "Command-Line Interface (CLI) for libcfnet.";
+
+static const char *const CF_NET_MANPAGE_LONG_DESCRIPTION =
+    "TODO - this man page is not written yet.";
+
+static const Description COMMANDS[] =
+{
+    {"help",    "Prints general help or per topic",
+                "cf-net help [command]"},
+    {"connect", "Checks if host(s) is available by connecting",
+                "cf-net -H 192.168.50.50,192.168.50.51 connect"},
+    {"stat",    "Look at type of file",
+                "cf-net stat masterfiles/update.cf"},
+    {"get",     "Get file from server",
+                "cf-net get masterfiles/update.cf -o download.cf"},
+    {"opendir", "List files and folders in a directory",
+                "cf-net opendir masterfiles"},
+    {NULL, NULL, NULL}
+};
+
+static const struct option OPTIONS[] =
+{
+    {"help",        no_argument,        0, 'h'},
+    {"manpage",     no_argument,        0, 'M'},
+    {"host",        required_argument,  0, 'H'},
+    {"debug",       no_argument,        0, 'd'},
+    {"verbose",     no_argument,        0, 'v'},
+    {"inform",      no_argument,        0, 'I'},
+    {NULL,          0,                  0, '\0'}
+};
+
+static const char *const HINTS[] =
+{
+    "Print the help message",
+    "Print the man page",
+    "Hostname or IP of server (defaults to policy_server.dat)",
```

Comment:
This can accept a comma separated list, am I right? It should probably be mentioned here.